### ISSUE TITLE
Excel: inconsistent application.calculationState in Online and Mac

### ISSUE THREAD

**Issue by wh1t3cAt1k** [OfficeDev/office-js#2883]
> Message from office-js bot:
We’re closing this issue because it has been inactive for a long time. We’re doing this to keep the issues list manageable and useful for everyone. If this issue is still relevant for you, please create a new issue.
> Thank you for your understanding and continued feedback.

## Your Environment

First environment:

- macOS monterey 12.3 21E230
- Excel for Mac 16.65(22091101)

Second environment:

- Excel for the Web
- Chrome browser latest

## Steps to reproduce

ScriptLab: https://gist.github.com/agazukin/cebcf4c5b184f10d608b73f7855505aa

1. Use the script lab code that is appended to the git gist
2. Click the run button on the UI of the "addin"
3. Write some numbers on the worksheet, and add a e.g. SUM() for those numbers.

## Current behavior

Excel for Mac: console shows `calculationState` as 'Done' but after a few seconds the status becomes 'Pending'
Excel on the Web: the status remains as 'Done' after calculation 
 
## Expected behavior

Consistent `calculationState` that eventually becomes and remains `Done` on all platforms (Mac, Online, Windows)

## Context

We encountered a Mac-specific bug in our add-in because we were depending on the fact that after `worksheetCollection.calculated` the status will eventually become (and remain) "Done". We worked around the issue but still would appreciate the consistent behaviour in the two platforms.

**Comment by lindalu-MSFT** [comment]
@AlexJerabek You're the only contact for Script Lab issues. Can you help triage this? Thanks, linda

**Comment by AlexJerabek** [comment]
@lindalu-MSFT This seems like an Excel for Mac issue. Assigning @JinghuiMS.

**Comment by jipinshi** [comment]
@wh1t3cAt1k ， not sure the scenario, in general, calculatestatus will have different status like pending and calculating and done. In general, users care about the status of done, when one calculation is done, it will trigger onCalculated event. Can you explain the scenario you need to take care of the status of calculate?

**Comment by wh1t3cAt1k** [comment]
@JinghuiMS @jipinshi 

To be completely clear, my logic **does run** after the `onCalculated` event. 

However, it takes some time and a couple of context syncs to get to the actual piece of code that needs to be executed, and by this time the calculation state might have already changed. My code needs not to run if the calculation state is not "done".

But it turns out that we cannot consistently depend on this property on different platforms. 

To look at this from another side, a valid question: why even expose a calculationState property in the API if it behaves completely differently between Mac and Online / Windows?

IMO just from this perspective this is a valid inconsistency bug and the Mac should behave the same as Online, i.e. eventually get to the "Done" state.

**Comment by jipinshi** [comment]
@wh1t3cAt1k, thanks for clarification. Will try to repro and open bug to track it.

**Comment by jipinshi** [comment]
@wh1t3cAt1k , After trying on MAC with 12.5.1 version and cannot repro the issue. If this is not blocking issue, we can focus on the issue blocking you go ahead firstly. Wonder what's number of the potential customers for the addin?

**Comment by wh1t3cAt1k** [comment]
@jipinshi not just potential customers, but we already have 1000 clients who we are looking forward to migrate from the legacy solution to the new office.js based solution, and we want to work together with you to make everything as robust as possible!

I realized why it was not reproducing for you - in parallel I had a second file where our add-in was installed, and the file had a lot of custom functions with complex dependencies, looks like once this file is refreshed, then calculation state gets "spoiled" for all other files opened on Mac in parallel.

I would like to get to the bottom of this - since it requires credentials, could I get your or someone's contact email to provide more detailed repro steps and credentials?

I am also open for a teams meeting at your convenience!

**Comment by wh1t3cAt1k** [comment]
@jipinshi the repro steps are provided below, I wish I had simpler steps, but this really needs to be reviewed on your end since you have access to the internals of Excel, and a better understanding of why the calculation state gets stuck in "Pending".

Video: https://youtu.be/lA274NJe8R4

The steps are all visible in the video, but here is the text description just in case:

- Use Excel for Mac desktop.
- Install Velixo add-in from the Microsoft Store: https://appsource.microsoft.com/en-us/product/office/wa200002311?tab=overview
- Install script lab too
- Open a new blank workbook.
- Insert the script lab snippet from the sample above (https://gist.github.com/agazukin/cebcf4c5b184f10d608b73f7855505aa).
- Run the snippet
- Click the "Run" button
- Add a =SUM(1,2,3) function to the workbook
- Open console of the script lab
- Observe repeating "done" state in the console
- Download and open this file **in parallel:** [RapidByte Acumatica demo (3).xlsx](https://github.com/OfficeDev/office-js/files/9768325/RapidByte.Acumatica.demo.3.xlsx)
- Click the Velixo button and wait until the "Demo" connection appears in the list. Click "Edit" pencil and change connection settings:
- URL: **https://test.velixo.com/2022R1**
- Creds: review / microsoft
- Click "Connect" button at the bottom of the editor. 
- !!! Warning: if the site is cold, the connection can fail after some time. Just repeat the attempt after 1 minute if it does.
- After connected, the connection editor will close. 
- Click the white "Refresh" button in the task pane (top left)
- The functions will go into "#BUSY". Wait until the green status message "Smart refresh complete"
- While it's retrieving data, periodically switch back to the first almost-blank workbook: observe that eventually the calculation state is forever stuck in "Pending"

In the video you can see that there are no cells with errors after refresh completed, and there seem to be no active functions waiting to be calculated. Still, the application is forever in "Pending" state, this definitely feels like a bug.

(even though the file calculation mode is Automatic!)

@jipinshi please help us get to the bottom of this!!!

I know this scenario is pretty complex but it's the best I can offer for now to make it 100% repro. I will be happy to jump on a call with you and assist with reproducing if needed. 

Thanks a lot

**Comment by jipinshi** [comment]
@wh1t3cAt1k , thanks for the details steps and currently I can repro the issue on MAC. do you need to depend on the other workbook's addin's running status and do something? Just trying to understand the scenario. If this is critical to you, is there workaround for you now? And is it possible to have more simple repro steps with two scriptlab addins? Not with current big data? The main purpose is that enable investigator easily to understand whether the customize functions has finished calculation when the calculate status is pending. And is it possible to identify which customize function lead the issue? If remove all customize function, does the problem still exist? It will help accelerate navigating down the issue with simple steps. Will also continue to investigate.

**Comment by wh1t3cAt1k** [comment]
@jipinshi is there a way to add custom functions with scriptlab? 

I would be grateful if you started looking as is, but if the above is possible I will work in parallel to prepare simple steps!

However this is the best I could come up with at the moment, I spent much time yesterday on that and could not find a much simpler setup.

It requires one-time setup but then reproduces really quickly once set up, every time:) I would be grateful if you could try with this example while I am searching for simpler steps.

Regarding criticality, the problem leads to infinite recalculations on Mac in some scenarios, we had to stub out the logic under some circumstances.

**Comment by jipinshi** [comment]
@wh1t3cAt1k , yes, scriptlab can use custom functions. you can check Functions next to run in the scriptlab ribbon. If can find which custom function lead to the issue may help this.

**Comment by wh1t3cAt1k** [comment]
@jipinshi do you have any visibility into the internal Excel status when it's in the pending state? What's it "waiting for"? If you have this information and could provide it on your end it can help us troubleshoot too

**Comment by jipinshi** [comment]
@wh1t3cAt1k ,  pending means Changes that trigger calculation have been made, but a recalculation has not yet been performed.
Have you ever consider to set app.calculationMode set manual to improve perf and may can avoid the issue. Here is the link [https://learn.microsoft.com/en-us/office/dev/add-ins/excel/performance](url)

**Comment by wh1t3cAt1k** [comment]
@jipinshi the meaning of the Pending status is clear, however what is not clear and looks like a bug is:

why in Excel Online the calculation state becomes Done eventually, as one would expect in Automatic calculation mode, and in Mac it gets stuck at Pending?

This is not right, something in Excel internal state is messed up at that moment, which is why I figured you guys should take a look at it, perhaps someone from the Excel for Mac team.

**Comment by wh1t3cAt1k** [comment]
@JinghuiMS @jipinshi the importance of this issue has just increased for our team, one of our key product functionalities ("Drilldown") does not work reliably on Excel for Mac because of the bug

**Comment by jipinshi** [comment]
@wh1t3cAt1k, I tried the scenario, and it will change to completed first and then go to pending after the log shows that all requests are done, I think there you can execute your logic in the callback. The first thing is trying to find some workaround to unblock this. As for the bug itself, still need your help to simple the steps in scriptlab to help dig down the root cause.

**Comment by wh1t3cAt1k** [comment]
@jipinshi so far I have not been able to simplify the steps. 😥

I was hoping for your insight regarding what is happening with Excel internally when it gets stuck in Pending — at MSFT you should have some sort of instruments for that, which are not available to external developers?

Is there some sort of "calculation queue" based on which Excel decides its state? If yes, is there anything in that queue that looks suspicious?..

I will continue trying to find a simpler setup but I would still ask you to try and investigate on the current setup (besides, except one-time setup costs it's really simple and fast to repro), there is no guarantee that I am able to do much better regarding the steps...

**Comment by wh1t3cAt1k** [comment]
Also, what is the log you're talking about that "all requests are done"? Can we catch this moment programmatically with minimal hacks?

**Comment by jipinshi** [comment]
@wh1t3cAt1k , internally use 6930899 to track the bug, will update once there is progress. Also want to confirm that does the issue blocking you since you mentioned that you have workaround for it at first.

Also want to confirm what did the addin do after run the addin, since on win32 I also noticed that the calculationstatus would change from finish to pending to finish and to pending, pending. It makes me feel that the addin is running continually.

**Comment by wh1t3cAt1k** [comment]
Yeah, for the original issue we did have a workaround, but now the same thing affects us in another place that is more important. It can be considered blocking.

Add-in doesn't have any kind of infinite logic, which is proved by the fact that on Excel Online all calculation will finish to Done eventually. 

What would be great is to understand why Excel is in Pending state, what it is waiting to recalculate!

**Comment by 4tti** [comment]
We are facing same issue unfortunately. We wanted to implement 'calculation in-progress' notification which would be visible for the end-user till all sheets are calculated but it is not possible, especially while onCalculated event is triggered per sheet and just for 'some' sheets (depending on formulas, etc.).

**Comment by wh1t3cAt1k** [comment]
Hi @jipinshi @RuizhiSunMS are there any news on the fixing or identifying the root cause? Any ETAs for the fix? This is a pretty pressing issue for us as some functionalities in the add-in simply don't work reliably on a Mac.

**Comment by JinghuiMS** [comment]
@wh1t3cAt1k 
We're actively working on the reported issues according to their priority. Currently we haven't had any ETA yet. We will share once we have it. Thanks!

**Comment by JinghuiMS** [comment]
@4tti
To better assist you, would you please show more details? Which platform you're seeing this error? What's your add-in's name? And the steps to repro this issue is also very important to us. Thanks!

**Comment by wh1t3cAt1k** [comment]
cc @Wenjun-Gong - this is one of the Velixo's "red" issues we discussed recently... it only reproduces on Excel Desktop for **Mac**, other platforms seem to reach the "Done" state eventually.

For the record, since it's a high impact production bug, to work around the issue, I am adding an "escape hatch" that if Excel is stuck for 5 seconds in a "Pending" state (and calculation mode is not manual), we consider that it is in fact "Done" and not pending.

Pseudocode of what we added:

```
let totalPendingStates = 0;

 do {
            context.application.load({
                calculationState: true,
                calculationMode: true,
            });

            await context.sync();

            if (
                context.application.calculationMode ===
                    Excel.CalculationMode.manual ||
                context.application.calculationState ===
                    Excel.CalculationState.done
            ) {
                logger.info('Calculation is finished. Exiting the check loop.');

                break;
            }

            if (
                context.application.calculationState ===
                Excel.CalculationState.pending
            ) {
                ++totalPendingStates;
            }

// !!!!!!!!!!!!!!!
// Start escape hatch
// !!!!!!!!!!!!!!!

            if (totalPendingStates === maximumPendingStates) {
                logger.info(
                    `Escape hatch: Excel seems to be stuck in Pending state for ${maximumPendingStates} seconds already, we observed this on Mac only. Will not wait for Done anymore.`
                );

                break;
            }

// !!!!!!!!!!!!!!!
// End escape hatch
// !!!!!!!!!!!!!!!

            logger.info(
                'Excel is still in calculation (we observed this on large files): waiting for 1 second before checking if calculation ended again.'
            );

            await delay(1000);
        } while (true);
        
// additional logic checking for active functions count and other stuff...
```

(this is incomplete logic for something similar that @4tti mentioned - we have quite a few operations where we need to have a high degree of confidence (I understand it can never be 100%, but good enough) that Excel has recomputed after our functions finished and will not "shoot back up again". For the record, just using "calculated" event is not reliable either because it can be triggered multiple times over the course of calculations)

I am sending the code above to illustrate how application being stuck in "Pending" can be harmful and lead to infinitely stuck operations on our end. It doesn't reproduce every time, but sporadically it does and results in support requests.

**Comment by Wenjun-Gong** [comment]
Hi Pavel, @JinghuiMS is investigating the root reason. Please stay tuned. Thanks.

**Comment by JinghuiMS** [comment]
@wh1t3cAt1k 
Happy new year! Sorry for the late response!

We'd like to confirm the below items for further investigation. Thanks in advance!
1. If Velixo can work fine with simple file in customer’s environment.
2. Contrary to what the attached video showed, the data in the worksheet was not successfully recalculated after we clicked the "Refresh All" button. If the re-calculation is not successful, the state should be pending. (BTW, whether the recent upgrade of Velixo is related to this issue)
![image](https://user-images.githubusercontent.com/56814588/210297020-f2fdd61f-0bc3-4410-9551-541934d9a651.png)
3. Which APIs are called by “refresh all” button

**Comment by 4tti** [comment]
> @4tti To better assist you, would you please show more details? Which platform you're seeing this error? What's your add-in's name? And the steps to repro this issue is also very important to us. Thanks!

Sorry for late response.
I personally saw it on Windows version, desktop Excel. 
The steps were easy - we got wrong state almost all the time when we try to calculate the report sheet by sheet. In the end we decided to show the progress for calculating only our custom functions (and we know when it ends).
If still interested we can try it once again.

**Comment by wh1t3cAt1k** [comment]
> @wh1t3cAt1k Happy new year! Sorry for the late response!
> 
> We'd like to confirm the below items for further investigation. Thanks in advance!
> 
> 1. If Velixo can work fine with simple file in customer’s environment.
> 2. Contrary to what the attached video showed, the data in the worksheet was not successfully recalculated after we clicked the "Refresh All" button. If the re-calculation is not successful, the state should be pending. (BTW, whether the recent upgrade of Velixo is related to this issue)
>    ![image](https://user-images.githubusercontent.com/56814588/210297020-f2fdd61f-0bc3-4410-9551-541934d9a651.png)
> 3. Which APIs are called by “refresh all” button

@JinghuiMS I feel like you're going through our own pain now 😅 The fact that you're seeing #NAME maybe relates to a case of add-in poisoning #2878 - I recommend to use "document inspector" to remove task pane add-ins from the file, and then inserting Velixo from the store. (and imagine we have to go through this with our real customers! Who pay for our software!)

On simple files, it never reproduces - we haven't been able to narrow down the root cause.

Answering your questions about the API, these are the ones we invoke:

- `context.workbook.dataConnections.refreshAll()`
- `context.workbook.application.calculate('Full')`
- `context.workbook.pivotTables.refreshAll()`

**Comment by JinghuiMS** [comment]
@wh1t3cAt1k 
Thanks for your information.
Let me try to use "document inspector" and re-install Velixo. Will get back to you if I have any new findings.

**Comment by wh1t3cAt1k** [comment]
Hi @JinghuiMS please let us know if there are any news about this bug

**Comment by wh1t3cAt1k** [comment]
@JinghuiMS @Wenjun-Gong dear friends, I would like to follow up on this issue to ask if there has been any development.

**Comment by JinghuiMS** [comment]
Sorry for the late reply, we just had a 7-day Chinese New Year holiday.
We have ENGs( +@donlvMSFT @SiruiSun-MSFT for awareness) are actively investigating the root cause and fixing approach. We will update this thread once we have fixed it Thanks!

**Comment by wh1t3cAt1k** [comment]
Hi @JinghuiMS @donlvMSFT @SiruiSun-MSFT , is there any news on the root cause and the fix ETA? Thank you!

**Comment by SiruiSun-MSFT** [comment]
Hi @wh1t3cAt1k ,
Sorry for the late reply, since this issue may be a bit complicated, we need more time to do a full investigation, if there is any progress, we will update here, thanks for your patience!

**Comment by SiruiSun-MSFT** [comment]
Closed by mistake, re-open this issue. And update progress: We are now working on the fix. Will release the fix to Prod ASAP. Thanks for your patience!

**Comment by SiruiSun-MSFT** [comment]
Thanks for waiting here! We are excited to share with you that this fix has been rolled out to Prod on version(16.86 on MAC and 16.0.17726 on Windows desktop). Thanks for your reports again!

**Comment by wh1t3cAt1k** [comment]
[Velixo internal tracking item](https://3.basecamp.com/3496385/buckets/16689259/card_tables/cards/7677685106)
